lib/kamikakushi/kaonashi.rb

```diff
@@ -0,0 +1,50 @@
+module Kamikakushi
+  module Kaonashi
+    extend ActiveSupport::Concern
+
+    module ClassMethods
+      def kaonashi(options = {})
+        define_singleton_method(:dependent_parent_name) { options[:parent] }
+        return unless dependent_parent_name
+
+        class_eval do
+          include InstanceMethods
+          default_scope { without_deleted }
+          alias_method_chain :destroyed?, :kaonashi
+
+          scope :with_deleted, -> {
+            join_with_dependent_parent(dependent_parent_name, :with_deleted)
+          }
+
+          scope :without_deleted, -> {
+            join_with_dependent_parent(dependent_parent_name, :without_deleted)
+          }
+
+          scope :only_deleted, -> {
+            join_with_dependent_parent(dependent_parent_name, :only_deleted)
+          }
+        end
+      end
+
+      private
+
+      def join_with_dependent_parent(dependent_parent_name, scope_name)
+        association =  reflect_on_all_associations.find { |a| a.name == dependent_parent_name }
+        dependent_parent_class =  association.klass
+
+        parent_arel = dependent_parent_class.arel_table
+        joins_conditions = arel_table.join(parent_arel)
+                                    .on(parent_arel[:id].eq arel_table[association.foreign_key])
+                                    .join_sources
+        joins(joins_conditions).merge(dependent_parent_class.__send__(scope_name))
```

Comment:
can't use `joins(association.name).merge(association.class.__send__(scope_name))` because [this issue](https://github.com/rails/rails/pull/18109)